uvicorn/_handlers/http.py

```diff
@@ -0,0 +1,68 @@
+import asyncio
+from typing import TYPE_CHECKING
+
+from uvicorn.config import Config
+
+if TYPE_CHECKING:  # pragma: no cover
+    from uvicorn.server import ServerState
+
+MAX_RECV = 65536
+
+
+async def handle_http(
+    reader: asyncio.StreamReader,
+    writer: asyncio.StreamWriter,
+    server_state: "ServerState",
+    config: Config,
+) -> None:
+    # Run transport/protocol session from streams.
+
+    # Use a future to coordinate between the protocol and this handler task.
+    # https://docs.python.org/3/library/asyncio-protocol.html#connecting-existing-sockets
+    loop = asyncio.get_event_loop()
+    connection_lost = loop.create_future()
+
+    # Switch the protocol from the stream reader to our own HTTP protocol class.
+    protocol = config.http_protocol_class(
+        config=config,
+        server_state=server_state,
+        on_connection_lost=lambda: connection_lost.set_result(True),
+    )
+    transport = writer.transport
+    transport.set_protocol(protocol)
+
+    # Asyncio stream servers don't `await` handler tasks, so attach a callback to
+    # retrieve and handle exceptions occurring in protocols outside the ASGI
+    # request-response cycle (e.g. bugs).
+    task = asyncio.current_task()
+    assert task is not None
+
+    @task.add_done_callback
+    def retrieve_exception(task: asyncio.Task) -> None:
+        exc = task.exception()
+        if exc is not None:
+            loop.call_exception_handler(
+                {
+                    "message": "Fatal error in server handler",
+                    "exception": exc,
+                    "transport": transport,
+                    "protocol": protocol,
+                }
+            )
+            # Hang up the connection so the client doesn't wait forever.
+            transport.close()
```

Comment:
I've tried this section by introducing a `RuntimeError` in `H11Protocol.send_events()` just before we kick the ASGI task.

* Without this low-level task handling, the exception is logged by asyncio as "never retrieved", and the client is left hanging because the transport doesn't get properly closed.
* With this low-level task handling, we properly retrieve the exception and close the transport, so the connection is hung up, the client notices, and handle the error. (This is in line with what's currently happening in `master`.) For example, with `curl` we get this output:

```console
$ curl localhost:8000
curl: (52) Empty reply from server
```